File: javascript/secret-handshake/secret-handshake.js
export const secretHandshake = (number) => {
  if (isNaN(number)) throw new Error('Handshake must be a number');

  const conditions = [ 
    { val: 0b1, text: 'wink' },
    { val: 0b10, text: 'double blink' },
    { val: 0b100, text: 'close your eyes' },
    { val: 0b1000, text: 'jump' },
  ];

  const isReverse = 0b10000;
  const seq = []

  conditions.forEach(c => {
    if (c.val & number) seq.push(c.text)
  });

  if (isReverse & number) seq.reverse();

  return seq;
};

File: go/secret-handshake/secret_handshake.go
package secret

var actions = [4]string{"wink", "double blink", "close your eyes", "jump"}

// Handshake converts a number into a sequence of actions for a secret handshake.
func Handshake(code uint) []string {
	result := make([]string, 0, len(actions))
	for i := 0; i < 4; i++ {
		if code&(1<<i) != 0 {
			result = append(result, actions[i])
		}
	}
	if code&(1<<4) != 0 {
		for i, j := 0, len(result)-1; i < j; i, j = i+1, j-1 {
			result[i], result[j] = result[j], result[i]
		}
	}
	return result
}
